src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,490 @@
+                    edge_types=edge_types_formatted,
+                    properties=properties_string,
+                )
+            )
+
+        object_domain = self._user_to_system_domain_map.get(
+            object_domain.lower(), object_domain
+        )
+
+        parent_param = ""
+        query_object_name = object_name
+        if object_version:
+            parent_param = f', {_ObjectField.PARENT_NAME}:"{object_name}"'
+            query_object_name = object_version
+
+        query_template = '{{{nodeKey}({domainKey}: {domain}, {nameKey}:"{name}"{parent_param}) {{{edges}}}}}'
+        query = query_template.format(
+            nodeKey=_DGQLFields.NODE,
+            domainKey=_ObjectField.DOMAIN,
+            nameKey=_ObjectField.NAME,
+            domain=object_domain.upper(),
+            name=query_object_name,
+            parent_param=parent_param,
+            edges="".join(parts),
+        )
+
+        return f"select SYSTEM$DGQL('{query}')"
+
+    def _get_lineage(
+        self,
+        object_name: str,
+        object_domain: str,
+        directions: list,
+        object_version: Optional[str] = None,
+        current_depth=1,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Constructs and executes a query to trace the lineage of a given entity at a single depth level.
+        """
+        query_string = self._build_graphql_query(
+            object_name, object_domain, directions, object_version
+        )
+        response = self._session.sql(query_string)
+        json_response = json.loads(response.collect()[0][0])
+
+        rows = []
+        for direction in directions:
+            edges = (
+                json_response.get(_DGQLFields.DATA, {})
+                .get(_DGQLFields.NODE, {})
+                .get(direction.value, [])
+            )
+
+            for edge in edges:
+                if _DGQLFields.SOURCE in edge and _DGQLFields.TARGET in edge:
+                    rows.append(
+                        (
+                            edge[_DGQLFields.SOURCE],
+                            edge[_DGQLFields.TARGET],
+                            direction,
+                            current_depth,
+                        )
+                    )
+
+        return rows
+
+    def _recursive_trace(
+        self,
+        object_name: str,
+        object_domain: str,
+        direction: str,
+        current_depth: int,
+        total_depth: int,
+        visited: set,
+        object_version: Optional[str] = None,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Recursively traces lineage by making successive DGQL queries based on response nodes.
+        """
+        if current_depth > total_depth:
+            return []
+
+        current_node = (object_name, object_domain, object_version)
+        if current_node in visited:
+            return []
+
+        visited.add(current_node)
```

Comment:
are you using one `visited` set for the entire graph search? I thought you need to use different `visited` for each path. What if you have 2 paths like
1. A->B->C->D
2. A-> D
3. B->D
If you visit D first in path 1, does that mean you cannot find 2nd and 3rd path now since D is visited?

Reply:
I see what you mean. I think BFS would be better to handle these cases. I have switched it to BFS. I have a visited set per direction.